# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmIonCompile.cpp
# Commit: aad37f6cbf03
# Full Hash: aad37f6cbf034f5ec3e0b04989aa025c7c899697
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-11 15:10:42
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.2: Establish dependency from tls for x86 callWithABI div/mod i64. r=lth
#   
#   x86 has few register so to do div/mod for i64 it call the runtime
#   and clobber almost all gp registers including WasmTlsReg.
#   To be able to call c++ runtime via Builtin thunk we need to set up WasmTlsReg.
# ==============================================================================

diff -r 05007ff436a5 -r aad37f6cbf03 js/src/wasm/WasmIonCompile.cpp
--- a/js/src/wasm/WasmIonCompile.cpp	Fri Sep 11 07:15:21 2020 +0000
+++ b/js/src/wasm/WasmIonCompile.cpp	Fri Sep 11 08:05:01 2020 +0000
@@ -460,6 +460,19 @@
       curBlock_->add(rhs2);
       rhs = rhs2;
     }
+
+    // For x86 we implement i64 div via c++ runtime.
+    // A call to c++ runtime requires tls pointer.
+#ifdef JS_CODEGEN_X86
+    if (type == MIRType::Int64) {
+      auto* ins =
+          MWasmBuiltinDivI64::New(alloc(), lhs, rhs, tlsPointer_, unsignd,
+                                  trapOnError, bytecodeOffset());
+      curBlock_->add(ins);
+      return ins;
+    }
+#endif
+
     auto* ins = MDiv::New(alloc(), lhs, rhs, type, unsignd, trapOnError,
                           bytecodeOffset(), mustPreserveNaN(type));
     curBlock_->add(ins);
@@ -481,6 +494,19 @@
       curBlock_->add(rhs2);
       rhs = rhs2;
     }
+
+    // This is because x86 codegen calls runtime via BuiltinThunk and so it
+    // needs WasmTlsReg to be live.
+#ifdef JS_CODEGEN_X86
+    if (type == MIRType::Int64) {
+      auto* ins =
+          MWasmBuiltinModI64::New(alloc(), lhs, rhs, tlsPointer_, unsignd,
+                                  trapOnError, bytecodeOffset());
+      curBlock_->add(ins);
+      return ins;
+    }
+#endif
+
     auto* ins = MMod::New(alloc(), lhs, rhs, type, unsignd, trapOnError,
                           bytecodeOffset());
     curBlock_->add(ins);
